src/lib/currency/currency.ml

```diff
@@ -195,6 +195,91 @@ end = struct
+            match kind with
+            | `Add ->
+                if Int.(Field.compare t (Lazy.force modulus_as_field) < 0) then
+                  (* Within range. *)
+                  Field.zero
+                else
+                  (* Overflowed. We compensate by subtracting [modulus_as_field]. *)
+                  Field.(negate one)
+            | `Sub ->
+                if Int.(Field.compare t (Lazy.force modulus_as_field) < 0) then
+                  (* Within range. *)
+                  Field.zero
+                else
+                  (* Underflowed, but appears as an overflow because of wrapping in
+                     the field (that is, -1 is the largest field element, -2 is the
+                     second largest, etc.). Compensate by adding [modulus_as_field].
+                  *)
+                  Field.one
+            | `Add_or_sub ->
+                (* This case is a little more nuanced: -modulus_as_field < t <
+                   2*modulus_as_field, and we need to detect which 'side of 0' we
+                   are. Thus, we have 3 cases:
+                *)
+                if Int.(Field.compare t (Lazy.force modulus_as_field) < 0) then
+                  (* 1. we are already in the desired range, no adjustment; *)
+                  Field.zero
+                else if
+                  Int.(Field.compare t (Lazy.force double_modulus_as_field) < 0)
+                then
+                  (* 2. we are in the range
+                        [modulus_as_field <= t < 2 * modulus_as_field],
+                        so this was an addition that overflowed, and we should
+                        compensate by subtracting [modulus_as_field];
+                  *)
+                  Field.(negate one)
+                else
+                  (* 3. we are outside of either range, so this must be the
+                        underflow of a subtraction, and we should compensate by
+                        adding [modulus_as_field].
+                  *)
+                  Field.one)
+    in
+    let%bind out_of_range =
+      match kind with
+      | `Add ->
+          (* 0 or -1 => 0 or 1 *)
+          Boolean.of_field (Field.Var.negate adjustment_factor)
+      | `Sub ->
+          (* Already 0 or 1 *)
+          Boolean.of_field adjustment_factor
+      | `Add_or_sub ->
+          (* The return flag [out_of_range] is a boolean represented by either 0
+             when [t] is in range or 1 when [t] is out-of-range.
+             Notice that [out_of_range = adjustment_factor^2] gives us exactly
+             the desired values, and moreover we can ensure that
+             [adjustment_factor] is exactly one of -1, 0 or 1 by checking that
+             [out_of_range] is boolean.
+          *)
+          Field.Checked.mul adjustment_factor adjustment_factor
+          >>= Boolean.of_field
+    in
+    (* [t_adjusted = t + adjustment_factor * modulus_as_field] *)
+    let t_adjusted =
```

Comment:
You should `seal` this here because the `Assert.equal` is going to anyway, so then you won't be duplicating the sealing in the call sites